src/python/pants/backend/jvm/subsystems/zinc.py

```diff
@@ -153,6 +214,108 @@ def compiler_interface(self):
     """
     return self._zinc_factory._compiler_interface(self._products)
 
+  @memoized_property
+  def scala_compiler(self):
+    """Return the path to the scala compiler jar.
+
+    :rtype: str
+    """
+    return self._zinc_factory._scala_compiler(self._products)
+
+  @memoized_property
+  def scala_library(self):
+    """Return the path to the scala library jar (runtime).
+
+    :rtype: str
+    """
+    return self._zinc_factory._scala_library(self._products)
+
+  @memoized_property
+  def scala_reflect(self):
+    """Return the path to the scala library jar (runtime).
+
+    :rtype: str
+    """
+    return self._zinc_factory._scala_reflect(self._products)
+
+  @memoized_property
+  def _compiler_bridge_cache_dir(self):
+    """A directory where we can store compiled copies of the `compiler-bridge`.
+
+    The compiler-bridge is specific to each scala version.
+    Currently we compile the `compiler-bridge` only once, while bootstrapping.
+    Then, we store it in the working directory under .pants.d/zinc/<cachekey>, where
+    <cachekey> is calculated using the locations of zinc, the compiler interface,
+    and the compiler bridge.
+    """
+    hasher = sha1()
+    for cp_entry in [self.zinc, self.compiler_interface, self.compiler_bridge]:
+      hasher.update(os.path.relpath(cp_entry, get_buildroot()))
+    key = hasher.hexdigest()[:12]
+
+    return os.path.join(self._zinc_factory.get_options().pants_workdir, 'zinc', 'compiler-bridge', key)
+
+  def _make_relative(self, path):
+    """A utility function to create relative paths to the work dir"""
+    return fast_relpath(path, get_buildroot())
+
+  def _run_bootstrapper(self, bridge_jar, context):
+    bootstrapper = self._zinc_factory._compiler_bootstrapper(self._products)
+    bootstrapper_args = [
+      "--out", bridge_jar,
+      "--compiler-interface", self.compiler_interface,
+      "--compiler-bridge-src", self.compiler_bridge,
+      "--scala-compiler", self.scala_compiler,
+      "--scala-library", self.scala_library,
+      "--scala-reflect", self.scala_reflect,
+    ]
+    input_jar_snapshots = context._scheduler.capture_snapshots((PathGlobsAndRoot(
+      PathGlobs(tuple([self._make_relative(jar) for jar in bootstrapper_args[1::2]])),
+      text_type(get_buildroot())
+    ),))
+    argv = tuple(['.jdk/bin/java'] +
+                 ['-cp', bootstrapper, Zinc.ZINC_BOOTSTRAPER_MAIN] +
+                 bootstrapper_args
+    )
+    req = ExecuteProcessRequest(
+      argv=argv,
+      input_files=input_jar_snapshots[0].directory_digest,
+      output_files=(self._make_relative(bridge_jar),),
+      output_directories=(self._make_relative(self._compiler_bridge_cache_dir),),
+      description="bootstrap compiler bridge.",
+      jdk_home=self.dist.home,
+    )
+    return context.execute_process_synchronously_or_raise(req, 'zinc-subsystem', [WorkUnitLabel.COMPILER])
+
+  def compile_compiler_bridge(self, context):
```

Comment:
This method is called for every zinc compile. Either the zinc task should memoize it, or this subsystem should. Following the example from #6491, I'd suggest memoizing it here by marking this method `@memoized`.